src/main/java/minijava/ir/emit/IREmitter.java

```diff
@@ -133,20 +142,29 @@ private void connectParametersToIRVariables(Method that) {
     int thisIdx = getLocalVarIndex(new LocalVariable(fakeThisType, null, SourceRange.FIRST_CHAR));
     // We rely on this when accessing this.
     assert thisIdx == 0;
+    Type fakeThisStorageType = storageType(fakeThisType);
     construction.setVariable(
-        thisIdx, construction.newProj(args, Types.storageType(fakeThisType).getMode(), thisIdx));
+        thisIdx, construction.newProj(args, fakeThisStorageType.getMode(), thisIdx));
+    NodeUtils.setLink(
+        construction.getVariable(thisIdx, fakeThisStorageType.getMode()), fakeThisStorageType.ptr);
 
     for (LocalVariable p : that.parameters) {
       // we just made this connection in the loop above
       // Also effectively this should just count up.
       // Also note that we are never trying to access this or the
       int idx = getLocalVarIndex(p);
       // Where is this documented anyway? SimpleIf seems to be the only reference...
-      Node param = construction.newProj(args, Types.storageType(p.type).getMode(), idx);
+      Type parameterType = storageType(p.type);
+      Node param = construction.newProj(args, parameterType.getMode(), idx);
+      NodeUtils.setLink(param, parameterType.ptr);
       construction.setVariable(idx, param);
     }
   }
 
+  private Type storageType(minijava.ast.Type type) {
+    return Types.storageType(type, classTypes);
+  }
```

Comment:
I'd prefer having a `Types` object that contains the `classTypes` rather than having static methods we pass all state into. Not really related to this PR, though.